File: legacy/backend_fastapi/app/models/routine.py
from datetime import datetime

from sqlalchemy import DateTime, ForeignKey, Integer, String, func
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.core.database import Base


class Routine(Base):
    __tablename__ = "routines"

    id: Mapped[int] = mapped_column(Integer, primary_key=True)
    user_id: Mapped[int] = mapped_column(
        ForeignKey("users.id"), nullable=False, index=True
    )
    name: Mapped[str] = mapped_column(String(150), nullable=False)
    goal: Mapped[str | None] = mapped_column(String(60), nullable=True)
    days_per_week: Mapped[int] = mapped_column(Integer, default=1)

    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now()
    )

    user: Mapped["User"] = relationship(back_populates="routines")
    days: Mapped[list["RoutineDay"]] = relationship(
        back_populates="routine",
        cascade="all, delete-orphan",
        order_by="RoutineDay.day_index",
    )


class RoutineDay(Base):
    __tablename__ = "routine_days"

    id: Mapped[int] = mapped_column(Integer, primary_key=True)
    routine_id: Mapped[int] = mapped_column(
        ForeignKey("routines.id"), nullable=False, index=True
    )
    day_index: Mapped[int] = mapped_column(Integer, nullable=False)
    name: Mapped[str] = mapped_column(String(150), nullable=False)
    muscle_focus: Mapped[str | None] = mapped_column(String(120), nullable=True)

    routine: Mapped["Routine"] = relationship(back_populates="days")
    exercises: Mapped[list["RoutineExercise"]] = relationship(
        back_populates="routine_day",
        cascade="all, delete-orphan",
        order_by="RoutineExercise.order",
    )


class RoutineExercise(Base):
    __tablename__ = "routine_exercises"

    id: Mapped[int] = mapped_column(Integer, primary_key=True)
    routine_day_id: Mapped[int] = mapped_column(
        ForeignKey("routine_days.id"), nullable=False, index=True
    )
    exercise_id: Mapped[int] = mapped_column(
        ForeignKey("exercises.id"), nullable=False, index=True
    )
    order: Mapped[int] = mapped_column(Integer, default=0)
    target_sets: Mapped[int] = mapped_column(Integer, default=3)
    target_reps_min: Mapped[int] = mapped_column(Integer, default=8)
    target_reps_max: Mapped[int] = mapped_column(Integer, default=12)
    target_rest_seconds: Mapped[int] = mapped_column(Integer, default=90)
    notes: Mapped[str | None] = mapped_column(String(500), nullable=True)

    routine_day: Mapped["RoutineDay"] = relationship(back_populates="exercises")
    exercise: Mapped["Exercise"] = relationship()

File: lib/models/routine.dart
import 'exercise.dart';

class RoutineExercise {
  final int id;
  final int order;
  final int targetSets;
  final int targetRepsMin;
  final int targetRepsMax;
  final int targetRestSeconds;
  final double? targetWeightKg;
  final String? notes;
  final ExerciseSummary exercise;

  RoutineExercise({
    required this.id,
    required this.order,
    required this.targetSets,
    required this.targetRepsMin,
    required this.targetRepsMax,
    required this.targetRestSeconds,
    this.targetWeightKg,
    this.notes,
    required this.exercise,
  });

  factory RoutineExercise.fromJson(Map<String, dynamic> json) =>
      RoutineExercise(
        id: json['id'],
        order: json['order'],
        targetSets: json['target_sets'],
        targetRepsMin: json['target_reps_min'],
        targetRepsMax: json['target_reps_max'],
        targetRestSeconds: json['target_rest_seconds'],
        targetWeightKg: (json['target_weight_kg'] as num?)?.toDouble(),
        notes: json['notes'],
        exercise: ExerciseSummary.fromJson(json['exercise']),
      );
}

class RoutineDay {
  final int id;
  final int dayIndex;
  final String name;
  final String? muscleFocus;
  final List<RoutineExercise> exercises;

  RoutineDay({
    required this.id,
    required this.dayIndex,
    required this.name,
    this.muscleFocus,
    required this.exercises,
  });

  factory RoutineDay.fromJson(Map<String, dynamic> json) => RoutineDay(
    id: json['id'],
    dayIndex: json['day_index'],
    name: json['name'],
    muscleFocus: json['muscle_focus'],
    exercises: (json['exercises'] as List)
        .map((e) => RoutineExercise.fromJson(e))
        .toList(),
  );
}

class Routine {
  final int id;
  final String name;
  final String? goal;
  final int daysPerWeek;
  final List<RoutineDay> days;

  Routine({
    required this.id,
    required this.name,
    this.goal,
    required this.daysPerWeek,
    required this.days,
  });

  factory Routine.fromJson(Map<String, dynamic> json) => Routine(
    id: json['id'],
    name: json['name'],
    goal: json['goal'],
    daysPerWeek: json['days_per_week'],
    days: (json['days'] as List).map((e) => RoutineDay.fromJson(e)).toList(),
  );
}

class RoutineSummary {
  final int id;
  final String name;
  final String? goal;
  final int daysPerWeek;
  final List<String> exerciseNames;

  RoutineSummary({
    required this.id,
    required this.name,
    this.goal,
    required this.daysPerWeek,
    this.exerciseNames = const [],
  });

  factory RoutineSummary.fromJson(Map<String, dynamic> json) => RoutineSummary(
    id: json['id'],
    name: json['name'],
    goal: json['goal'],
    daysPerWeek: json['days_per_week'],
  );
}
